### PR TITLE
Optimizations for hard equations

### PR DESCRIPTION
**Pull request by vhavlena**
Optimizations for hard equations

- extending Nielsen transformation to handle general equations (no termination guarantee though)
- length-based pruning of the proof graph
- both side trimming of the nodes
- various other fixes

### CODE REVIEW

**Comment by vhavlena** [comment]
@jurajsic Before merging, could you please check if the `Kepler` benchmark is ok? I made some changes that could possibly affect this benchmark.

**Comment by jurajsic** [comment]
image
image
On `woorpje` we still have 80 TO/MEMOUT (see attached .txt), but much better than 1.0.0, which had 243. On Kepler we have 3 TO, which is 1 less.
[message.txt](https://github.com/VeriFIT/z3-noodler/files/13383229/message.txt)
[message(1).txt](https://github.com/VeriFIT/z3-noodler/files/13383231/message.1.txt)

**Comment by vhavlena** [comment]
Thanks a lot! The results look nice. We can get back and try to solve the remaining 80 TOs at some point.

**Review by Adda0 — COMMENTED** [review]
Seems fine to me overall. I will approve later when I check a few more parts more thoroughly.

File: src/smt/theory_str_noodler/nielsen_decision_procedure.h
Comment (vhavlena) [reply src/smt/theory_str_noodler/nielsen_decision_procedure.h]: You are right. I fixed the comment.

**Review by jurajsic — APPROVED** [review]
Looks fine, I can't say I fully understand it though.

**Comment by vhavlena** [comment]
@Adda0 Can you finish your review? I would like to merge it to move on.

**Review by Adda0 — APPROVED** [review]
I reviewed the rest of the changes. LGTM.

**Comment by vhavlena** [comment]
Great, I merge it today.

**Pull request merged**
